### PR TITLE
MDEV-16029 mysqldump: dump and restore historical data

### LINKED ISSUE #505: MDEV-16029 mysqldump: dump and restore historical data
### Description
Backup through mysqldump loses historical data from versioned tables.

mysqldump does not read historical rows from the versioned tables, it runs SELECT, not SELECT .. FOR SYSTEM_TIME ALL. But even if we fixed it, it would still be impossible to load this data back, since we are dealing either with invisible columns into which we can't insert at all, or with virtual columns for which values are ignored. So, for now it becomes a known limitation which needs to be documented.

### TODO
- [x] check if table is versioned via `SHOW CREATE`;
- [x] INSERT statements for history;
- [x] test case;
- [x] Test explicit system fields;
- [x] Test modes:
- [x] opt_compact
- [x] path (--tab)
- [x] --no-create-info
- [x] other codeflow branches?
- [x] option to dump with history or current snapshot only;
- [x] Warn about opt_xml;
- [ ] (optionally) dump at specific point in time [MDEV-16355].

### Links
* https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_sql_quote_show_create
* https://mariadb.com/kb/en/library/comment-syntax/
* https://dev.mysql.com/doc/refman/5.5/en/load-xml.html
* https://stackoverflow.com/questions/2934258/how-do-i-get-the-current-time-zone-of-mysql
* https://dev.mysql.com/doc/refman/8.0/en/time-zone-support.html
* https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html

### PR DESCRIPTION
**Pull request by midenok**
--dump-history option includes history into table's dumped data.

Also fixes
MDEV-16733 mysqldump --tab and --xml options are conflicting

[fixes tempesta-tech/mariadb#505]

### CODE REVIEW

File: mysql-test/suite/versioning/t/data.test [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+if (0)
+{
+--echo # XML with history
+drop table t1;
+create or replace table t1 (x int) with system versioning;
+delete from t2;
+delete history from t2;
+set system_versioning_modify_history= ON;
+--replace_result $TMP TMP
+eval load xml infile '$TMP/dump_history.xml' into table t1;
```
Comment (FooBarrior) [inline mysql-test/suite/versioning/t/data.test:84]: Better to call it `load.test`, not `data.test`

File: client/mysqldump.c [hunk 2]
```diff
@@ -2842,24 +2865,23 @@ static uint get_table_structure(char *table, char *db, char *table_type,
...
+    if (switch_character_set_results(mysql, "binary") ||
+        mysql_query_with_error_report(mysql, &result, buff) ||
+        switch_character_set_results(mysql, default_charset))
+    {
+      my_free(order_by);
+      order_by= 0;
+      DBUG_RETURN(0);
+    }
+
     if (!opt_no_create_info)
```
Comment (FooBarrior) [inline client/mysqldump.c:2880]: please invert this if-else, to avoid double inversion

File: client/mysqldump.c [hunk 3]
```diff
@@ -3100,10 +3145,33 @@ static uint get_table_structure(char *table, char *db, char *table_type,
...
+            mysql_query_with_error_report(mysql, &result, query_buff) ||
+            switch_character_set_results(mysql, default_charset))
+        DBUG_RETURN(0);
+
+      row= mysql_fetch_row(result);
+
+      if (strstr(row[1], "WITH SYSTEM VERSIONING"))
+      {
+        *versioned= 1;
+        if (0 == strstr(row[1], "GENERATED ALWAYS AS ROW START"))
```
Comment (FooBarrior) [inline client/mysqldump.c]: why is it "hidden"? Shouldn't You check `INVISIBLE` instead?

**Review by FooBarrior — CHANGES_REQUESTED** [review]

File: client/mysqldump.c (hunk 2 above)
Comment (midenok) [reply client/mysqldump.c:2880]: What does double inversion mean?

File: mysql-test/suite/versioning/t/data.test (hunk 1 above)
Comment (midenok) [reply mysql-test/suite/versioning/t/data.test:84]: Test is not only about load. Nor it not only about backup/restore.

File: client/mysqldump.c (hunk 2 above)
Comment (FooBarrior) [reply client/mysqldump.c:2880]: It turns out, that it's called [Reverse Conditional](https://www.refactoring.com/catalog/reverseConditional.html). Double negative is [here](https://refactoring.com/catalog/removeDoubleNegative.html).

TLDR `if (opt_no_create_info){...} else {...}` will be more readable

Comment (midenok) [reply client/mysqldump.c:2880]: No, in fact it won't. The main block which was there initially is more important. The second block is minor case.

**Review by FooBarrior — APPROVED** [review]
